### DISCUSSION TITLE
How to set TranslatorInterface locale based on User entity in stateless firewall

### DISCUSSION THREAD

**Discussion by crtl** [symfony/symfony#49024] [Q&A] [answered]
In symfony you are supposed to set the Requests locale before an internal event listener configures the `TranslatorInterface`.

> The custom listener must be called before LocaleListener, which initializes the locale based on the current request. To do so, set your listener priority to a higher value than LocaleListener priority (which you can obtain by running the debug:event kernel.request command).

https://symfony.com/doc/current/translation/locale.html

My problem is that I have to access the users locale but the listeners run on `kernel.request` event which is executed before the firewall (ie user is not logged in at this moment).
How am I supposed to configure the translator based on users locale stored in database?

I have tried the solution from this thread (https://stackoverflow.com/questions/57995513/locale-based-on-the-users-preferences) but its not working.
I can only set the requests locale but the translator is not picking it up.

Currently as a workaround I have to manually pass the `locale` argument to all `TranslatorInterface::trans` calls.

**EDIT**
Calling `setLocale` on `TranslatorInterface` works but is obviously no solution because TranslatorInterface has no method `setLocale`

**Comment by stof** [comment, marked as answer]
The solution is to use the LocaleSwitcher service to change the locale if running after the LocaleListener: https://symfony.com/doc/current/translation.html#switch-locale-programmatically

**Reply by crtl** [reply]
Thank you very much.
